equinox/nn/conv.py

```diff
@@ -0,0 +1,227 @@
+from ..module import Module, static_field
+
+
+def _ntuple(n):
+    def parse(x):
+        if isinstance(x, collections.abc.Iterable):
+            return tuple(x)
+        return tuple(repeat(x, n))
+
+    return parse
+
+
+class Conv(Module):
+    num_spatial_dims: int = static_field()
+    weight: Array
+    bias: Optional[Array]
+    in_channels: int = static_field()
+    out_channels: int = static_field()
+    kernel_size: Union[int, Sequence[int]] = static_field()
+    stride: Union[int, Sequence[int]] = static_field()
+    padding: Union[int, Sequence[int]] = static_field()
+    dilation: Union[int, Sequence[int]] = static_field()
+    use_bias: bool = static_field()
+
+    def __init__(
+        self,
+        num_spatial_dims,
+        in_channels,
+        out_channels,
+        kernel_size,
+        stride=1,
+        padding=0,
+        dilation=1,
+        use_bias=True,
+        *,
+        key,
+        **kwargs,
+    ):
+        super().__init__()
+        self.num_spatial_dims = num_spatial_dims
+        parse = _ntuple(self.num_spatial_dims)
+        wkey, bkey = jrandom.split(key, 2)
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.kernel_size = parse(kernel_size)
+        self.use_bias = use_bias
+        lim = 1 / math.sqrt(self.in_channels * math.prod(self.kernel_size))
+
+        self.weight = jrandom.uniform(
+            wkey,
+            (
+                self.out_channels,
+                self.in_channels,
+            )
+            + self.kernel_size,
+            minval=-lim,
+            maxval=lim,
+        )
+        if self.use_bias:
+            self.bias = jrandom.uniform(
+                bkey,
+                (self.out_channels,) + (1,) * self.num_spatial_dims,
+                minval=-lim,
+                maxval=lim,
+            )
+        else:
+            self.bias = None
+
+        self.stride = parse(stride)
+        if type(padding) == tuple and len(padding) == self.num_spatial_dims:
+            self.padding = tuple(
+                [parse(padding[i]) for i in range(self.num_spatial_dims)]
+            )
+        elif type(padding) == int:
+            self.padding = tuple(
+                [(padding, padding) for _ in range(self.num_spatial_dims)]
+            )
+        else:
+            raise ValueError(
+                f"Padding must either be int or tuple of length {self.num_spatial_dims}"
+            )
+        self.dilation = parse(dilation)
+
+    def __call__(self, x, *, key=None):
+        unbatched_rank = self.num_spatial_dims + 1
+        allowed_ranks = [unbatched_rank, unbatched_rank + 1]
+        if x.ndim not in allowed_ranks:
+            raise ValueError(
+                f"Input to Conv needs to have rank in {allowed_ranks},",
+                f" but input has shape {x.shape}.",
+            )
+        unbatched = x.ndim == unbatched_rank
+        if unbatched:
+            x = jnp.expand_dims(x, axis=0)
```

Comment:
So I'm tempted to suggest that only the unbatched case should be supported here. (Wth the appropriate un/squeezing.) WDYT?

Reply:
Sounds good to me - probably nice to enforce using vmap for more consistent style